### PR TITLE
feat(crons): Create MonitorIncident objects

### LINKED ISSUE #56580: Add logic to create/update objects
Create/resolve `MonitorIncident` objects in concert with `mark_failed` and `mark_ok`

### PR DESCRIPTION
**Pull request by rjo100**
Creates + resolves `MonitorIncident` objects in accordance with thresholds.

Closes  https://github.com/getsentry/sentry/issues/56580

### CODE REVIEW

File: src/sentry/monitors/models.py [hunk 1]
```diff
@@ -583,6 +584,15 @@ def get_last_successful_checkin(self):
...
         )
 
+    def get_incident_grouphash(self):
+        fingerprint = [
+            "monitor",
+            str(self.monitor.guid),
+            self.environment.name,
+            str(self.last_state_change),
+        ]
+        return hash_from_values(fingerprint)
```
Comment (evanpurkhiser) [inline src/sentry/monitors/models.py:594]:
```suggestion
    @property
    def incident_grouphash(self):
        fingerprint = [
            "monitor",
            str(self.monitor.guid),
            self.environment.name,
            str(self.last_state_change),
        ]
        return hash_from_values(fingerprint)
```

maybe this?

Comment (evanpurkhiser) [reply src/sentry/monitors/models.py:594]: Wait why do we even need to do this? We can just generate a unique UUID per incident? the incident row is the source of truth

File: src/sentry/monitors/logic/mark_failed.py [hunk 2]
```diff
@@ -98,6 +121,8 @@ def mark_failed_threshold(
             .order_by("-date_added")
             .first()
         ]
+
+        fingerprint = monitor_env.get_incident_grouphash()
```
Comment (rjo100) [inline src/sentry/monitors/logic/mark_failed.py:125]: need to add a test case for this still

File: src/sentry/monitors/models.py [hunk 3]
```diff
@@ -583,6 +584,30 @@ def get_last_successful_checkin(self):
...
+            fingerprint = hash_from_values(
+                [
+                    "monitor",
+                    str(self.monitor.guid),
+                    self.environment.name,
+                    str(self.last_state_change),
+                ]
+            )
+
+        return fingerprint
```
Comment (evanpurkhiser) [inline src/sentry/monitors/models.py:609]: mostly a nit
```suggestion
        if active_incident:
        		return active_incident.grouphash
        
        # XXX(rjo100): While we migrate monitor issues to using the 
        # Incident stored grouphash we still may have some active issues 
        # that are using the old hashes. We can remove this in the 
        # future once all existing issues are resolve.
        return hash_from_values(
            [
                "monitor",
                str(self.monitor.guid),
                self.environment.name,
                str(self.last_state_change),
            ]
        )
```

File: src/sentry/monitors/models.py [hunk 4]
```diff
@@ -604,11 +629,11 @@ class MonitorIncident(Model):
     monitor_environment = FlexibleForeignKey("sentry.MonitorEnvironment")
     starting_checkin = FlexibleForeignKey(
         "sentry.MonitorCheckIn", null=True, related_name="created_incidents"
-    )
+    )  # this represents the first failed check-in that we receive
     starting_timestamp = models.DateTimeField(null=True)
```
Comment (evanpurkhiser) [inline src/sentry/monitors/models.py:633]: Let's write this as a doc string

```suggestion
    )
    starting_timestamp = models.DateTimeField(null=True)
    """
    This represents the first failed check-in that we receive
    """

```

File: src/sentry/monitors/models.py [hunk 5]
```diff
@@ -604,11 +629,11 @@ class MonitorIncident(Model):
...
     resolving_checkin = FlexibleForeignKey(
         "sentry.MonitorCheckIn", null=True, related_name="resolved_incidents"
-    )
+    )  # this represents the final OK check-in that we receive
     resolving_timestamp = models.DateTimeField(null=True)
```
Comment (evanpurkhiser) [inline src/sentry/monitors/models.py:637]:
```suggestion
    )
    resolving_timestamp = models.DateTimeField(null=True)
    """
    This represents the final OK check-in that we receive
    """

```

**Review by evanpurkhiser — APPROVED** [review]

**Pull request merged**